presto-cache/src/main/java/com/facebook/presto/cache/UnboundedCacheManager.java

```diff
@@ -0,0 +1,163 @@
+
+    private final Path directory;
+
+    public UnboundedCacheManager(Path directory)
+    {
+        this.directory = requireNonNull(directory, "directory is null");
+
+        File target = new File(directory.toString());
+        if (!target.exists()) {
+            target.mkdir();
+        }
+        else {
+            File[] files = target.listFiles();
+            if (files == null) {
+                return;
+            }
+            Arrays.stream(files).forEach(File::delete);
+        }
+    }
+
+    @Override
+    public boolean get(FileCacheKey key, byte[] buffer, int offset)
+    {
+        drainInflights();
+
+        Slice data = inflight.get(key);
+        if (data != null) {
+            checkState(key.getLength() == data.length(), "cache length mismatch %s vs %s", key.getLength(), data.length());
+            checkState(
+                    offset + key.getLength() <= buffer.length,
+                    "insufficient buffer size %s; requested [%s, %s)",
+                    buffer.length,
+                    offset,
+                    offset + key.getLength());
+
+            data.getBytes(0, buffer, offset, key.getLength());
+            return true;
+        }
+
+        Path path = persisted.get(key);
+        if (path != null) {
+            try (FileInputStream input = new FileInputStream(path.toString())) {
+                input.read(buffer, offset, key.getLength());
+                return true;
+            }
+            catch (IOException e) {
+                return false;
+            }
+        }
+
+        return false;
+    }
+
+    @Override
+    public void put(FileCacheKey key, Slice data)
+    {
+        drainInflights();
+
+        inflight.put(key, data);
+        try {
+            executor.submit(() -> {
+                Path path = new Path(directory.toString() + "/" + randomUUID() + EXTENSION);
+                if (!persist(key, data, path)) {
+                    log.warn("Fail to persist cache %s with length %s ", path, key.getLength());
+                }
+
+                inflight.remove(key);
+            }).get();
+        }
+        catch (Exception e) {
+            e.printStackTrace();
+        }
+    }
+
+    private void drainInflights()
+    {
+        while (!inflightToPersistedJobs.isEmpty()) {
+            Future<?> job = inflightToPersistedJobs.poll();
+
+            // in case another thread has polled the last job
+            if (job == null) {
+                return;
+            }
+
+            if (!job.isDone()) {
+                // return back to the queue
+                inflightToPersistedJobs.add(job);
+
+                // ignore the rest of the jobs given it's unlikely they have finished.
+                return;
+            }
+        }
+    }
+
+    private boolean persist(FileCacheKey key, Slice data, Path path)
+    {
+        try {
+            Files.copy(data.getInput(), (new File(path.toString())).toPath(), REPLACE_EXISTING);
```

Comment:
If the input `FileCacheKey` is only a part of the file, it will rewrite what has been cached.